climada_petals/hazard/rf_glofas/river_flood_computation.py

```diff
@@ -0,0 +1,658 @@
+
+        Parameters
+        ----------
+        store_intermediates : bool, optional
+            Whether the data of each computation step should be stored in the cache
+            directory. This is recommended especially for larger data. Only set this
+            to ``False`` if the data operated on is very small (e.g., for a small
+            country or region). Defaults to ``True``.
+        data_dir : Path or str, optional
+            The directory where flood maps, Gumbel distribution parameters and the
+            FLOPROS database are located. Defaults to
+            ``<climada>/data/river-flood-computation``, where ``<climada>`` is the
+            Climada data directory indicated by ``local_data : system`` in the
+            ``climada.conf``. This directory must exist.
+        cache_dir : Path or str, optional
+            The top-level cache directory where computation caches of this instance will
+            be placed. Defaults to ``<climada>/data/river-flood-computation/.cache``
+            (see above for configuration). This directory (and all its parents) will be
+            created.
+        """
+        data_dir = Path(data_dir)
+        if not data_dir.is_dir():
+            raise FileNotFoundError(f"'data_dir' does not exist: {data_dir}")
+
+        self.store_intermediates = store_intermediates
+        self.flood_maps = xr.open_dataarray(
+            data_dir / "flood-maps.nc",
+            chunks=dict(return_period=-1, latitude="auto", longitude="auto"),
+        )
+        self.gumbel_fits = xr.open_dataset(data_dir / "gumbel-fit.nc", chunks="auto")
+        self.flopros = gpd.read_file(data_dir / "FLOPROS_shp_V1/FLOPROS_shp_V1.shp")
+        self.regridder = None
+        self._create_tempdir(cache_dir=cache_dir)
+
+    def _create_tempdir(self, cache_dir: Union[Path, str]):
+        """Create a temporary directory inside the top-level cache dir
+
+        Parameters
+        ----------
+        cache_dir : Path or str
+            The directory where caches are placed. Each cache is a temporary
+            subdirectory of ``cache_dir``. If the path does not exist, it will be
+            created, including all parent directories.
+        """
+        # Create cache directory
+        cache_dir = Path(cache_dir)
+        cache_dir.mkdir(parents=True, exist_ok=True)
+
+        # Create temporary directory for cache
+        self._tempdir = TemporaryDirectory(
+            dir=cache_dir, prefix=datetime.today().strftime("%y%m%d-%H%M%S-")
+        )
+
+        # Define cache paths
+        self.cache_paths = RiverFloodCachePaths.from_dir(Path(self._tempdir.name))
+
+    def clear_cache(self):
+        """Clear the cache of this instance
+
+        This will delete the temporary cache directory and create a new one by calling
+        :py:meth:`_create_tempdir`.
+        """
+        cache_dir = Path(self._tempdir.name).parent
+        self._tempdir.cleanup()
+        self._create_tempdir(cache_dir=cache_dir)
```

Comment:
I tried to find an elegant solution to this. I personally favor `TemporaryDirectory` over `mkdtemp` because it actually *tries* to clean up implicitly. But I see that I should not rely on it exclusively. So here's my proposal:

* Use `TemporaryDirectory`
* Call `cleanup` when I intend to clean up the directory before re-assigning the `_tempdir`.
* Add a `__del__` method which also calls `cleanup` to avoid implicit cleanup, although this [seems discouraged](https://stackoverflow.com/a/2452895/22190523).

I'll also add the appropriate calls to `cleanup` to the tests

Reply:
That sounds a bit like you are reinventing the wheel. Maybe just copy what `TemporaryDirectory` does internally and use `weakref.finalize`: https://github.com/JakobDev/cpython/blob/92ac7ba794271f9437993b859799dc6f4355b1f8/Lib/tempfile.py#L869

Reply:
Don't you think re-implementing internals of `TemporaryDirectory` would be reinventing the wheel? I think the code remains much more understandable when we stick to `TemporaryDirectory`. Please have a look at the latest version, I think it should be both safe and understandable.

Reply:
@tovogt How to continue here? I think the current implementation using `__del__` is simple and clean.

Reply:
I still think, it's not unsatisfactory that using `__del__` cannot guarantee to suppress the warnings by `TemporaryDirectory`. But admittedly, using weakref feels clumsy since you have to define a callback and keep references of it around etc. So, I agree that `__del__` is simple, but I don't think it's clean. Still, I agree to keep it like it is.